tests/lib/state_test.py

```diff
@@ -138,3 +140,122 @@ def _extra_state(self):
                 {'name': 'eth1', 'state': 'up', 'type': 'unknown'}
             ]
         })
+
+
+class TestAssertRouteState(object):
+    IPV4_ROUTE1_IFACE_NAME = 'eth1'
+
+    IPV4_ROUTE1 = {
+        Route.DESTINATION: '198.51.100.0/24',
+        Route.METRIC: 102,
+        Route.NEXT_HOP_ADDRESS: '192.0.2.2',
+        Route.NEXT_HOP_INTERFACE: IPV4_ROUTE1_IFACE_NAME,
+        Route.TABLE_ID: 254
+    }
+
+    IPV4_ROUTE2_IFACE_NAME = 'eth2'
+
+    IPV4_ROUTE2 = {
+        Route.DESTINATION: '203.0.113.0/24',
+        Route.METRIC: 102,
+        Route.NEXT_HOP_ADDRESS: '192.0.2.2',
+        Route.NEXT_HOP_INTERFACE: IPV4_ROUTE2_IFACE_NAME,
+        Route.TABLE_ID: 254
+    }
+
+    def test_merge_route_add(self):
+        """
+        Test adding routes with old routes preserved.
+        """
+        current_state = state.State(self._base_state_dict)
+        desired_state_dict = copy.deepcopy(self._base_state_dict)
+        new_route = copy.deepcopy(TestAssertRouteState.IPV4_ROUTE1)
+        new_route[Route.NEXT_HOP_ADDRESS] = '192.0.2.3'
+        desired_state_dict[Route.KEY][Route.CONFIG] = [new_route]
+        desired_state = state.State(desired_state_dict)
+        desired_state.merge_route_config(current_state)
+
+        assert new_route in desired_state.iface_routes['eth1']
+        iface_routes = desired_state.iface_routes
+        # Old routes should be no touched
+        assert (TestAssertRouteState.IPV4_ROUTE1 in
+                iface_routes[TestAssertRouteState.IPV4_ROUTE1_IFACE_NAME])
+        # Interface without changes should not be included
+        assert (TestAssertRouteState.IPV4_ROUTE2 not in
+                iface_routes[TestAssertRouteState.IPV4_ROUTE2_IFACE_NAME])
+
+    def test_merge_route_add_without_iface(self):
+        """
+        Test desired_state only contains routes without any interface defined.
+        """
+        current_state = state.State(self._base_state_dict)
+        desired_state_dict = copy.deepcopy(self._base_state_dict)
+        new_route = copy.deepcopy(TestAssertRouteState.IPV4_ROUTE1)
+        new_route[Route.NEXT_HOP_ADDRESS] = '192.0.2.3'
+        desired_state_dict[Route.KEY][Route.CONFIG] = [new_route]
+        desired_state_dict[Interface.KEY] = {}
+        desired_state = state.State(desired_state_dict)
+        changed_iface_names = desired_state.merge_route_config(current_state)
+        desired_state.include_changed_interfaces(current_state,
+                                                 changed_iface_names)
+
+        assert (new_route[Route.NEXT_HOP_INTERFACE] in
+                desired_state.interfaces)
+
+    def test_verify_routes_add_route(self):
```

Comment:
How about: `test_routes_verification_ignores_order`.

In the test body, please reverse the list instead of using a tmp.
You can use an immutable ([::-1]) or mutable (list.reverse()) version.